BUG: oauth2client deprecated, use google-auth instead.

tswast opened the pull request:
Remove the use of oauth2client and use google-auth library, instead.

Rather than check for multiple versions of the libraries, use the
setup.py to specify compatible versions. I believe this is safe since
Pandas checks for the pandas_gbq package.

Since google-auth does not use the argparse module to override user
authentication flow settings, add a parameter to choose between the web
and console flow.

Closes https://github.com/pydata/pandas-gbq/issues/37.

tswast commented:
Travis build w/ integration tests: https://travis-ci.org/tswast/pandas-gbq/builds/233446493

/cc @jonparrott

jreback commented on ci/requirements-2.7.pip:
```diff
@@ -1,5 +1,7 @@
-python-gflags==2.0
-oauth2client==1.5.0
+httplib2>=0.9.1
+google-api-python-client>=1.6.2, <2.0.0dev
```
is there a reason you are putting the ``<2.0.0.dev``, ?

tswast replied:
Yes, because if there is a 2.0 release by semantic versioning that is a breaking change.

jreback replied:
thanks is not common practice and just hides errors

tswast replied:
Sure, I can drop it.

jreback commented on ci/requirements-3.4.pip:
```diff
@@ -1,3 +1,5 @@
-httplib2
-google-api-python-client
-oauth2client
+httplib2>=0.9.1
```
the idea of the different pip files would be to different versions, so for sure these should NOT have an upper bound

parthea replied:
Can we remove `httplib2` from `setup.py` altogether and use `from googleapiclient.http import build_http` in the code?


```
http = httplib2.Http()
```
becomes
```
http = build_http()
```

parthea replied:
Ok, Thanks for clarifying.

jreback commented on docs/source/changelog.rst:
```diff
@@ -4,7 +4,8 @@ Changelog
 
-- Resolve issue where the optional ``--noauth_local_webserver`` command line argument would not be propagated during the authentication process.
+- Use the `google-auth <https://google-auth.readthedocs.io/en/latest/>`__ library for authentication because oauth2client is deprecated. :issue:`37`
+- ``read_gbq`` now has a ``auth_local_webserver`` boolean argument for controlling whether to use web server or console flow when getting user credentials.
```
use the PR number as the issue number and (use parens)

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -6,57 +6,19 @@
 import time
 import sys
 
+import google.auth
```
no, you have to have a ``_try_imports`` to guard against these missing. The parent package (pandas) *only* checks that gbq itself can be imported, w/o a helpful message (hey you are missing a dependency).

tswast replied:
But I included google-auth in the dependencies in setup.py? What case would the try imports protect against?

gbq is an optional dependency, but once you install it, pip will install all of its dependencies.

jreback replied:
because u get unintelligible import exceptions otherwise

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -188,78 +172,103 @@ def get_application_default_credentials(self):
+            credentials do not have access to the project (self.project_id)
+            on BigQuery.
+        """
+        with open('bigquery_credentials.dat') as credentials_file:
```
this will hard fail with an odd message if this file doesn't exist. is there a reason this is hardcoded? (not to mention pathed), better to have its location specified by argument.

tswast replied:
The file path was already hard-coded with the previous `oauth2client.fileStorage('bigquery_credentials.dat')` call.

I've filed https://github.com/pydata/pandas-gbq/issues/41 to track adding a parameter to override this. Since this would be a new feature, I'd prefer to do this in a separate pull request.

I have added a check for if this file is unable to be opened/parsed plus an integration test for this case.

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -188,78 +172,103 @@ def get_application_default_credentials(self):
+        """
+        Saves user account credentials to a local file.
+        """
+        with open('bigquery_credentials.dat', 'wb') as credentials_file:
```
same

tswast replied:
This method will now log a message if the file cannot be saved. Since this doesn't prevent the credentials being used, it no longer throws an exception.

I've added a test for the case.

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -188,78 +172,103 @@ def get_application_default_credentials(self):
-            client_secret='kOc9wMptUtxkcIFbtZCcrEAc',
-            scope=self.scope,
-            redirect_uri='urn:ietf:wg:oauth:2.0:oob')
+    def get_user_account_credentials(self, auth_local_webserver=True):
```
are you expecting a user to call this method directly? should this not be in the read_/to_ API?

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -188,78 +172,103 @@ def get_application_default_credentials(self):
+
+        client_config = {
+            'installed': {
+                'client_id': ('495642085510-k0tmvj2m941jhre2nbqka17vqpjfddtd'
```
why is this hardcoded?

tswast replied:
Because it was hard coded previously. I just converted the variable to the expected format. This config variable is how Google knows to say Pandas GBQ is requesting access versus some other project.

jreback commented on pandas_gbq/tests/test_gbq.py:
```diff
@@ -83,100 +83,7 @@ def _get_private_key_contents():
         return f.read()
 
 
-def _test_imports():
```
pls don't remove things like this. just change to use the new deps. This is going to hard fail if deps are not installed.

jreback replied:
furthermore the point of this is to have informative messages.

jreback commented on setup.py:
```diff
@@ -17,9 +17,14 @@ def readme():
+    'pandas',
+    'httplib2>=0.9.1',
+    'google-api-python-client>=1.6.2, <2.0.0dev',
+    'google-auth>=1.0.0, <2.0.0dev',
```
you need to verify whether these dep packages are avail on conda-forge as well (and if not, prod to have them built there). furthermore unless there is a really really good reason don't put an upper bound on deps.

parthea commented:
@tswast Thanks for the PR! I'm going to try to test this tonight. I'm hitting this error in my local testing. Have you encountered it before?

```
    def setup_method(self, method):
        _setup_common()
        _skip_if_no_project_id()
        _skip_local_auth_if_in_travis_env()
    
>       self.sut = gbq.GbqConnector(_get_project_id())

test_gbq.py:180: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../gbq.py:120: in __init__
    auth_local_webserver=auth_local_webserver)
../gbq.py:128: in get_credentials
    credentials = self.get_application_default_credentials()
../gbq.py:175: in get_application_default_credentials
    credentials, _ = google.auth.default(scopes=[self.scope])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

scopes = ['https://www.googleapis.com/auth/bigquery'], request = None

    def default(scopes=None, request=None):
        """Gets the default credentials for the current environment.
    
        `Application Default Credentials`_ provides an easy way to obtain
        credentials to call Google APIs for server-to-server or local applications.
        This function acquires credentials from the environment in the following
        order:
    
        1. If the environment variable ``GOOGLE_APPLICATION_CREDENTIALS`` is set
           to the path of a valid service account JSON private key file, then it is
           loaded and returned. The project ID returned is the project ID defined
           in the service account file if available (some older files do not
           contain project ID information).
        2. If the `Google Cloud SDK`_ is installed and has application default
           credentials set they are loaded and returned.
    
           To enable application default credentials with the Cloud SDK run::
    
                gcloud auth application-default login
    
           If the Cloud SDK has an active project, the project ID is returned. The
           active project can be set using::
    
                gcloud config set project
    
        3. If the application is running in the `App Engine standard environment`_
           then the credentials and project ID from the `App Identity Service`_
           are used.
        4. If the application is running in `Compute Engine`_ or the
           `App Engine flexible environment`_ then the credentials and project ID
           are obtained from the `Metadata Service`_.
        5. If no credentials are found,
           :class:`~google.auth.exceptions.DefaultCredentialsError` will be raised.
    
        .. _Application Default Credentials: https://developers.google.com\
                /identity/protocols/application-default-credentials
        .. _Google Cloud SDK: https://cloud.google.com/sdk
        .. _App Engine standard environment: https://cloud.google.com/appengine
        .. _App Identity Service: https://cloud.google.com/appengine/docs/python\
                /appidentity/
        .. _Compute Engine: https://cloud.google.com/compute
        .. _App Engine flexible environment: https://cloud.google.com\
                /appengine/flexible
        .. _Metadata Service: https://cloud.google.com/compute/docs\
                /storing-retrieving-metadata
    
        Example::
    
            import google.auth
    
            credentials, project_id = google.auth.default()
    
        Args:
            scopes (Sequence[str]): The list of scopes for the credentials. If
                specified, the credentials will automatically be scoped if
                necessary.
            request (google.auth.transport.Request): An object used to make
                HTTP requests. This is used to detect whether the application
                is running on Compute Engine. If not specified, then it will
                use the standard library http client to make requests.
    
        Returns:
            Tuple[~google.auth.credentials.Credentials, Optional[str]]:
                the current environment's credentials and project ID. Project ID
                may be None, which indicates that the Project ID could not be
                ascertained from the environment.
    
        Raises:
            ~google.auth.exceptions.DefaultCredentialsError:
                If no credentials were found, or if the credentials found were
                invalid.
        """
        from google.auth.credentials import with_scopes_if_required
    
        explicit_project_id = os.environ.get(
            environment_vars.PROJECT,
            os.environ.get(environment_vars.LEGACY_PROJECT))
    
        checkers = (
            _get_explicit_environ_credentials,
            _get_gcloud_sdk_credentials,
            _get_gae_credentials,
            lambda: _get_gce_credentials(request))
    
        for checker in checkers:
            credentials, project_id = checker()
            if credentials is not None:
                credentials = with_scopes_if_required(credentials, scopes)
                return credentials, explicit_project_id or project_id
    
>       raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
E       DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or
E       explicitly create credential and re-run the application. For more
E       information, please see
E       https://developers.google.com/accounts/docs/application-default-credentials.

../../../miniconda2/lib/python2.7/site-packages/google/auth/_default.py:282: DefaultCredentialsError
```

tswast commented:
You're right, I forgot to check for if the token file exists, and I didn't fail gracefully if application default credentials were not set.

I'll add some tests for those cases tomorrow.

parthea commented:
@tswast At your earliest convenience, please could you rebase to resolve conflicts? Also, please run `flake8 pandas_gbq -v` to resolve flake8 errors.

```
pandas_gbq/gbq.py:46:9: F401 'google.auth' imported but unused
pandas_gbq/gbq.py:46:27: E261 at least two spaces before inline comment
pandas_gbq/gbq.py:46:28: E262 inline comment should start with '# '
pandas_gbq/gbq.py:47:9: F401 'google_auth_oauthlib.flow.InstalledAppFlow' imported but unused
pandas_gbq/gbq.py:47:63: E261 at least two spaces before inline comment
pandas_gbq/gbq.py:47:64: E262 inline comment should start with '# '
pandas_gbq/gbq.py:48:9: F401 'google_auth_httplib2' imported but unused
pandas_gbq/gbq.py:48:36: E261 at least two spaces before inline comment
pandas_gbq/gbq.py:48:37: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:92:41: E241 multiple spaces after ','
pandas_gbq/tests/test_gbq.py:120:13: F401 'google.auth.default' imported but unused
pandas_gbq/tests/test_gbq.py:120:44: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:120:45: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:121:13: F401 'google.auth.exceptions.DefaultCredentialsError' imported but unused
pandas_gbq/tests/test_gbq.py:121:71: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:121:72: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:122:62: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:122:63: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:123:13: F811 redefinition of unused 'Credentials' from line 122
pandas_gbq/tests/test_gbq.py:123:13: F401 'google.oauth2.service_account.Credentials' imported but unused
pandas_gbq/tests/test_gbq.py:123:66: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:123:67: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:129:13: F401 'google_auth_httplib2.AuthorizedHttp' imported but unused
pandas_gbq/tests/test_gbq.py:129:60: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:129:61: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:130:13: F401 'google_auth_httplib2.Request' imported but unused
pandas_gbq/tests/test_gbq.py:130:53: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:130:54: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:136:13: F401 'google_auth_oauthlib.flow.InstalledAppFlow' imported but unused
pandas_gbq/tests/test_gbq.py:136:67: E261 at least two spaces before inline comment
pandas_gbq/tests/test_gbq.py:136:68: E262 inline comment should start with '# '
pandas_gbq/tests/test_gbq.py:221:57: F841 local variable 'e' is assigned to but never used
pandas_gbq/tests/test_gbq.py:232:17: E129 visually indented line with same indent as next logical line
pandas_gbq/tests/test_gbq.py:296:32: F841 local variable 'auth_mock' is assigned to but never used
```

parthea commented:
@tswast It's great to see that most integration tests are passing with the improvements made in this PR. Do you have any thoughts on how to resolve the following integration test failure? 

https://travis-ci.org/tswast/pandas-gbq/jobs/233838815#L1293

parthea commented:
Could we change 
```
self.sut = gbq.GbqConnector(_get_project_id())
```
to
```
        self.sut = gbq.GbqConnector(_get_project_id(),
                                    auth_local_webserver=True)
```
in `setup_method()` in the `TestGBQConnectorIntegrationWithLocalUserAccountAuth` class to keep the functionality of the integration test consistent with the current master?

parthea commented:
I received a `resourceInUse` exception in my local testing. I believe these exceptions appeared before intermittently but they will appear less frequently with your improvements. Do you have any suggestions on how to further reduce the number of `resourceInUse` exceptions?

```
test_gbq.py::TestToGBQIntegrationWithServiceAccountKeyPath::test_upload_data_if_table_exists_fail ERROR
================================= short test summary info ==================================
SKIP [1] /home/tony/pydata-pandas-gbq/pandas_gbq/tests/test_gbq.py:304: Cannot get default_credentials from the environment!

========================================== ERRORS ==========================================
 ERROR at teardown of TestToGBQIntegrationWithServiceAccountKeyPath.test_upload_data_if_table_exists_fail 

self = <pandas_gbq.tests.test_gbq.TestToGBQIntegrationWithServiceAccountKeyPath object at 0x7f9915c5dad0>
method = <bound method TestToGBQIntegrationWithServiceAccountKeyPath.test_upload_data_i...st_gbq.TestToGBQIntegrationWithServiceAccountKeyPath object at 0x7f9915c5dad0>>

    def teardown_method(self, method):
        # - PER-TEST FIXTURES -
        # put here any instructions you want to be run *AFTER* *EVERY* test is
        # executed.
>       clean_gbq_environment(self.dataset_prefix, _get_private_key_path())

test_gbq.py:1033: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dataset_prefix = 'pandas_gbq_99300', private_key = '.....'

    def clean_gbq_environment(dataset_prefix, private_key=None):
        dataset = gbq._Dataset(_get_project_id(), private_key=private_key)
        all_datasets = dataset.datasets()
    
        retry = 3
        while retry > 0:
            try:
                retry = retry - 1
                for i in range(1, 10):
                    dataset_id = dataset_prefix + str(i)
                    if dataset_id in all_datasets:
                        table = gbq._Table(_get_project_id(), dataset_id,
                                           private_key=private_key)
                        all_tables = dataset.tables(dataset_id)
                        for table_id in all_tables:
                            try:
                                table.delete(table_id)
                            except gbq.NotFoundException as e:
                                pass
    
                        dataset.delete(dataset_id)
                retry = 0
            except gbq.GenericGBQException as ex:
                # Build in retry logic to work around the following errors :
                # An internal error occurred and the request could not be...
                # Dataset ... is still in use
                error_message = str(ex).lower()
                if ('an internal error occurred' in error_message or
                    'still in use' in error_message) and retry > 0:
                    sleep(randint(1, 10))
                else:
>                   raise ex
E                   GenericGBQException: Reason: resourceInUse, Message: Dataset pandas-140401:pandas_gbq_993001 is still in use

test_gbq.py:235: GenericGBQException
!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!
```

parthea commented:
If I set my `GOOGLE_APPLICATION_CREDENTIALS` environment variable to an empty string, I receive an exception instead of the local user web authentication flow. Is it expected?
```
tony@tonypc:~/pydata-pandas-gbq/pandas_gbq/tests$ export GOOGLE_APPLICATION_CREDENTIALS=''
tony@tonypc:~/pydata-pandas-gbq/pandas_gbq/tests$ pytest test_gbq.py::TestGBQConnectorIntegrationWithLocalUserAccountAuth -v -r s . --maxfail=1
============================= test session starts ==============================
platform linux2 -- Python 2.7.12, pytest-3.0.6, py-1.4.32, pluggy-0.4.0 -- /home/tony/miniconda2/bin/python
cachedir: ../../.cache
rootdir: /home/tony/pydata-pandas-gbq, inifile: 
plugins: cov-2.4.0
collected 100 items 

test_gbq.py::TestGBQConnectorIntegrationWithLocalUserAccountAuth::test_should_be_able_to_make_a_connector ERROR

==================================== ERRORS ====================================
 ERROR at setup of TestGBQConnectorIntegrationWithLocalUserAccountAuth.test_should_be_able_to_make_a_connector 

self = <pandas_gbq.tests.test_gbq.TestGBQConnectorIntegrationWithLocalUserAccountAuth object at 0x7f2bbecf6490>
method = <bound method TestGBQConnectorIntegrationWithLocalUserAccountAuth.test_should_....TestGBQConnectorIntegrationWithLocalUserAccountAuth object at 0x7f2bbecf6490>>

    def setup_method(self, method):
        _setup_common()
        _skip_if_no_project_id()
        _skip_local_auth_if_in_travis_env()
    
>       self.sut = gbq.GbqConnector(_get_project_id(), auth_local_webserver=True)

test_gbq.py:268: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../gbq.py:174: in __init__
    auth_local_webserver=auth_local_webserver)
../gbq.py:182: in get_credentials
    credentials = self.get_application_default_credentials()
../gbq.py:212: in get_application_default_credentials
    credentials, _ = google.auth.default(scopes=[self.scope])
../../../miniconda2/lib/python2.7/site-packages/google/auth/_default.py:277: in default
    credentials, project_id = checker()
../../../miniconda2/lib/python2.7/site-packages/google/auth/_default.py:138: in _get_explicit_environ_credentials
    os.environ[environment_vars.CREDENTIALS])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

filename = ''

    def _load_credentials_from_file(filename):
        """Loads credentials from a file.
    
        The credentials file must be a service account key or stored authorized
        user credentials.
    
        Args:
            filename (str): The full path to the credentials file.
    
        Returns:
            Tuple[google.auth.credentials.Credentials, Optional[str]]: Loaded
                credentials and the project ID. Authorized user credentials do not
                have the project ID information.
    
        Raises:
            google.auth.exceptions.DefaultCredentialsError: if the file is in the
                wrong format.
        """
>       with io.open(filename, 'r') as file_obj:
E       IOError: [Errno 2] No such file or directory: ''

../../../miniconda2/lib/python2.7/site-packages/google/auth/_default.py:63: IOError
!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
=========================== 1 error in 0.61 seconds ============================
```

tswast commented:
Yes, we should set the auth server to true in setup as you suggest. It's a good default to be false so it works in remote Jupyter notebooks, but py.test doesn't like it due to the output capturing.

I think resource in use errors are happening because the table deletes are eventually consistent. I will add a loop to list the tables until they are gone.

The GOOGLE_APPLICATION_CREDENTIALS set to empty string error sounds like a bug in google-auth. Let's file an issue over there for it.

tswast commented:
Okay, I've rebased, modified the integration tests to use local webserver by default, and added a couple more checks for invalid credentials.

Integration tests are running here: https://travis-ci.org/tswast/pandas-gbq/builds/234115597

tswast commented:
Looks like all the test failures are in the dataset listing

```
GenericGBQException: Reason: notFound, Message: Not found: Token pandas_gbq_457494
```

I'll ping the BQ team that this is still an issue. Not sure how to work around that one, since it seems pretty clearly a backend bug (we're getting a token from the API, but then they can't find it when we use it)

tswast commented:
Oh, and latest pandas removed assertRaises.

jreback commented on ci/requirements-2.7-0.19.2.pip:
```diff
@@ -1,5 +1,7 @@
-google-api-python-client==1.2
-python-gflags==2.0
-oauth2client==1.5.0
+google-api-python-client
```
can you fix one of these builts to the oldest versions that are possible to make this work (of google-api-python-client at least)

jreback commented on docs/source/changelog.rst:
```diff
@@ -4,8 +4,9 @@ Changelog
-- Resolve issue where the optional ``--noauth_local_webserver`` command line argument would not be propagated during the authentication process. (:issue:`35`)
 - Drop support for Python 3.4 (:issue:`40`)
+- Use the `google-auth <https://google-auth.readthedocs.io/en/latest/>`__ library for authentication because oauth2client is deprecated. (:issue:`39`)
+- ``read_gbq`` now has a ``auth_local_webserver`` boolean argument for controlling whether to use web server or console flow when getting user credentials.
```
is this a replacement for 35? if so, pls list that here

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -21,10 +21,9 @@ def _check_google_client_version():
-        google_api_minimum_version = '1.2.0'
+    # Version 1.6.0 is the first version to support google-auth.
+    # https://github.com/google/google-api-python-client/blob/master/CHANGELOG
+    google_api_minimum_version = '1.6.0'
```
right, so put this version fixed in the ci

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -188,78 +204,130 @@ def get_application_default_credentials(self):
+        return _try_credentials(self.project_id, credentials)
+
+    def save_user_account_credentials(self, credentials):
+        """
```
for new functions, add a ``versionadded 0.2.0`` in the doc-string

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -188,78 +204,130 @@ def get_application_default_credentials(self):
+        return credentials
 
+    def get_service_account_credentials(self):
+        import httplib2
```
are any of these new functions user API visible? if so, do a _try_imports at the top of them.

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -686,6 +750,9 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
+    auth_local_webserver : boolean (default False)
+        Use a local webserver when getting user credentials to handle
+        OAuth authorization flow redirects.
 
```
versionadded tag

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -686,6 +750,9 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
         Service account private key in JSON format. Can be file path
         or string contents. This is useful for remote server
         authentication (eg. jupyter iPython notebook on remote host)
+    auth_local_webserver : boolean (default False)
```
boolean, default False

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -686,6 +750,9 @@ def read_gbq(query, project_id=None, index_col=None, col_order=None,
         or string contents. This is useful for remote server
         authentication (eg. jupyter iPython notebook on remote host)
+    auth_local_webserver : boolean (default False)
+        Use a local webserver when getting user credentials to handle
```
if you can add a doc-link link on how this works would be great (in addition or just to gbq docs is fine as well)

jreback commented on pandas_gbq/gbq.py:
```diff
@@ -815,6 +883,9 @@ def to_gbq(dataframe, destination_table, project_id, chunksize=10000,
         authentication (eg. jupyter iPython notebook on remote host)
+    auth_local_webserver : boolean (default False)
+        Use a local webserver when getting user credentials to handle
+        OAuth authorization flow redirects.
```
versionadded tag & doc-link

jonparrott commented on pandas_gbq/gbq.py:
```diff
@@ -42,19 +41,50 @@ def _test_google_api_imports():
+        from googleapiclient.errors import HttpError  # noqa
+    except ImportError as ex:
+        raise ImportError(
+            "pandas requires google-api-python-client for Google  BigQuery "
```
Is there a stray space between `Google` and `BigQuery`?

tswast replied:
Yes, removed. Thanks.

jonparrott commented on pandas_gbq/gbq.py:
```diff
@@ -42,19 +41,50 @@ def _test_google_api_imports():
+    except ImportError as ex:
+        raise ImportError(
+            "pandas requires google-api-python-client for Google  BigQuery "
+            "support: {0}".format(str(ex)))
```
When using `.format`, you don't need to `str`, `format` will call `__str__`.

tswast replied:
Removed.

jonparrott commented on pandas_gbq/gbq.py:
```diff
@@ -42,19 +41,50 @@ def _test_google_api_imports():
+
+    http = httplib2.Http()
+    try:
+        http = AuthorizedHttp(credentials, http=http)
```
nit: shadowing variables makes me uncomfortable, I'd prefer you call this one `authed_http`.

tswast replied:
Done.

tswast commented:
Okay to merge?

parthea commented:
I'm really sorry for the lack of communication on my part. I want to get this merged asap but I've had limited time at the computer outside of work due to my newborn arriving this month (we're actually expecting in the next 12 hours!). I am planning to test the code once more before merging but unfortunately I cannot complete the testing until the weekend. @jreback may be able to merge this sooner (or possibly temporarily grant you repo write permission) but I don't feel comfortable merging this without completing basic tests. I will be off work for a few weeks once the baby is born and plan to spend time contributing to this project.

tswast commented:
Oh! Yeah, no rush. Family comes first.

parthea commented:
Unrelated to the changes in this PR,  I had a thought that we could temporarily add default credentials in the Travis environment in order to exercise `GoogleCredentials.get_application_default()` on Travis (created issue #49) as a separate PR.

parthea commented:
Unrelated to this PR, I noticed that the user experience when GBQ credentials are expired or revoked could be improved. I receive a `RefreshError` exception when I test revoked credentials. The user has to manually trigger the authentication flow. We could provide users with an option to automatically trigger the authentication flow if Google credentials are expired or revoked. I've created issue #50 to track this.

```
test_gbq.py::TestGBQConnectorIntegrationWithLocalUserAccountAuth::test_should_be_able_to_make_a_connector ERROR

=================================================================================== ERRORS ====================================================================================
________________________________ ERROR at setup of TestGBQConnectorIntegrationWithLocalUserAccountAuth.test_should_be_able_to_make_a_connector ________________________________

self = <pandas_gbq.tests.test_gbq.TestGBQConnectorIntegrationWithLocalUserAccountAuth object at 0x7fba2db9b590>
method = <bound method TestGBQConnectorIntegrationWithLocalUserAccountAuth.test_should_....TestGBQConnectorIntegrationWithLocalUserAccountAuth object at 0x7fba2db9b590>>

    def setup_method(self, method):
        _setup_common()
        _skip_if_no_project_id()
        _skip_local_auth_if_in_travis_env()
    
        self.sut = gbq.GbqConnector(
>           _get_project_id(), auth_local_webserver=True)

test_gbq.py:187: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../gbq.py:202: in __init__
    self.credentials = self.get_credentials()
../gbq.py:212: in get_credentials
    credentials = self.get_user_account_credentials()
../gbq.py:329: in get_user_account_credentials
    credentials = self.load_user_account_credentials()
../gbq.py:287: in load_user_account_credentials
    credentials.refresh(request)
../../../miniconda2/lib/python2.7/site-packages/google/oauth2/credentials.py:126: in refresh
    self._client_secret))
../../../miniconda2/lib/python2.7/site-packages/google/oauth2/_client.py:189: in refresh_grant
    response_data = _token_endpoint_request(request, token_uri, body)
../../../miniconda2/lib/python2.7/site-packages/google/oauth2/_client.py:109: in _token_endpoint_request
    _handle_error_response(response_body)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

response_body = '{
  "error" : "invalid_grant",
  "error_description" : "Token has been expired or revoked."
}'

    def _handle_error_response(response_body):
        """"Translates an error response into an exception.
    
        Args:
            response_body (str): The decoded response data.
    
        Raises:
            google.auth.exceptions.RefreshError
        """
        try:
            error_data = json.loads(response_body)
            error_details = '{}: {}'.format(
                error_data['error'],
                error_data.get('error_description'))
        # If no details could be extracted, use the response data.
        except (KeyError, ValueError):
            error_details = response_body
    
        raise exceptions.RefreshError(
>           error_details, response_body)
E       RefreshError: ('invalid_grant: Token has been expired or revoked.', u'{\n  "error" : "invalid_grant",\n  "error_description" : "Token has been expired or revoked."\n}')

../../../miniconda2/lib/python2.7/site-packages/google/oauth2/_client.py:59: RefreshError
```

parthea commented:
I received the following warning when running `flake8` on branch `tswast:gbq-37` but Travis did not report the warning (issue #51). 

Local run:
```
tony@tonypc:~/pydata-pandas-gbq$ flake8 pandas_gbq 
pandas_gbq/tests/test_gbq.py:136:61: F841 local variable 'e' is assigned to but never used
```

On Travis:
https://travis-ci.org/tswast/pandas-gbq/jobs/234914390

tswast commented:
🤦‍♂️ Thanks, good catch. Weirdly I only get it when I run flake8 with Python 2.7 but not on Python 3.6.

parthea commented:
Thanks for the patch, and for clarifying that this warning is for python 2.7 only (Travis is only running `flake8` on python 3.6). In the future, we could switch to using tox to make testing easier in local/Travis environment (Created issue #52).

parthea commented:
@jreback  Do we need to update https://github.com/conda-forge/pandas-gbq-feedstock/blob/master/recipe/meta.yaml to include the new deps?

jreback commented on ci/requirements-2.7-0.19.2.pip:
```diff
@@ -1,5 +1,7 @@
-httplib2
-google-api-python-client==1.2
-python-gflags==2.0
```
instead of fixing these on 2.7, pls do so on 3.5

tswast replied:
Done.

jreback commented on setup.py:
```diff
@@ -17,9 +17,14 @@ def readme():
+    'httplib2>=0.9.2',
+    'google-api-python-client>=1.6.0',
+    'google-auth>=1.0.0',
+    'google-auth-httplib2>=0.0.1',
```
none of the google-auth* libraries are available on conda-forge. So the way forward is to form recipices for this (or better yet to get the google team to do this).

parthea replied:
I could try to tackle this using the the [conda-forge FAQ](https://github.com/conda-forge/staged-recipes#faq) and examples. I'm not familiar with conda-forge so this would allow me to get up to speed faster. In the first version I'm going to add you both as maintainers in the conda-forge recipe for google-auth @tswast @jreback , but please let me know if you'd like to be removed.

tswast replied:
Thanks @parthea!

jreback commented:
that's fine 

u might want to look at how the google-api-python-client recipie looks

jreback commented:
ok this lgtm. let's see how the deps building on conda-forge do.

jreback commented:
@tswast ok I think need to rebase on master. still waiting for conda-forge deps to be accepted (shouldn't be long).

parthea commented:
Rebase would be helpful but as long as there are no conflicts with the base branch I'm happy to rebase when merging while maintaining authorship. @tswast Unfortunately, I didn't do a good job of maintaining authorship for your 2 commits in #58. That was my mistake. I will try really hard to maintain authorship when merging in the future.

jreback commented:
@parthea we use the button nowadays. As long as no conflicts this can be a ff merge. In any event this is blocked until the conda packages are available and its tested there.

parthea commented:
I agree. The Github merge button is my first choice. I'll stay on top of the conda package PRs so we can get this PR merged asap.

tswast commented:
Thanks for patching those in. I've rebased on master.

parthea commented:
@jreback I'm going to merge this now. All review comments have been addressed and integration tests passed.

tswast commented:
Awesome. Thanks @parthea.

parthea commented:
[google-auth](https://github.com/conda-forge/google-auth-feedstock), [google-auth-oauthlib](https://github.com/conda-forge/google-auth-oauthlib-feedstock) and [google-auth-httplib2](https://github.com/conda-forge/google-auth-httplib2-feedstock) are all conda packages.

jonparrott commented:
@parthea thanks for helping us get on conda!

parthea commented:
@jonparrott No problem. It was a great learning experience! Would you like to be added to the maintainer's list for the `google-auth*` conda packages?

jonparrott commented:
Sure.

On Fri, Jun 16, 2017, 9:38 AM Anthonios Partheniou <notifications@github.com>
wrote:

> @jonparrott <https://github.com/jonparrott> No problem. It was a great
> learning experience! Would you like to be added to the maintainer's list
> for the google-auth* conda packages?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pydata/pandas-gbq/pull/39#issuecomment-309074213>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAPUc_kgIR-7ym19g4q_PMVyuyoaPG6vks5sEq9ugaJpZM4Nelh1>
> .
>

jreback commented:
manually check the installation to make sure it's picking things up

may need to update install ibstructions as well

jreback commented:
ok need to change one of the builds to install all the deps via conda rather than pip

maybe the 3.6 build

volunteers?

tswast commented:
You mean in Travis? I've been meaning to learn Conda, so I'll volunteer.

jreback commented:
yep

parthea commented:
The installation of google-auth* conda packages appear to be working correctly with the following commands:

```
conda install -c conda-forge google-auth
conda install -c conda-forge google-auth-oauthlib
conda install -c conda-forge google-auth-httplib2
```

Yes, we'll need to update the documentation as well. I'll open a new issue.

tworec commented:
great job guys!
